File: scala-module/src/main/scala/scalabook/ImperativeNQueens.scala
package scalabook

trait NQueens {

}

class ImperativeNQueens(val size: Int) extends NQueens {

  val chess = Array.ofDim[Int](size)
  val Empty = -1
  init

  def isPossible(level: Int, i: Int): Boolean = {
    for (j <- 0 until level) {
      if (chess(j) == i
        || (level - j) == Math.abs(i - chess(j))) {
        return false
      }
    }
    true
  }

  private def init() {
    for (i <- 0 until size)
      chess(i) = Empty
  }

  def printSolutions() {
    def loop(level: Int) {
      if (level == size) {
        printResult
      } else {
        for (i <- 0 until size) {
          if (isPossible(level, i)) {
            chess(level) = i
            loop(level + 1)
            chess(level) = Empty
          }
        }
      }
    }

    loop(0)
  }

  private def printResult = {
    println(chess.mkString("#", ",", "#"))
  }
}

class FunctionalNQueens extends NQueens {


}

object NQueensTest {
  def main(args: Array[String]): Unit = {
    val nQueens = new ImperativeNQueens(5)
    nQueens.printSolutions()
  }
}

File: java-module/src/main/java/pattern/asyncmethod/ImperativeNQueens.java
package pattern.asyncmethod;

import java.util.Arrays;

public class ImperativeNQueens {

    private int[] queens;
    private final int SIZE;

    private static final int EMPTY = -1;

    public ImperativeNQueens(int size) {
        queens = new int[size];
        this.SIZE = size;
        Arrays.fill(queens, EMPTY);
    }

    public static void main(String[] args) {
        ImperativeNQueens nQueens = new ImperativeNQueens(5);
        nQueens.printSolutions();
    }

    void printSolutions() {
        putQueen(0);
    }

    private void putQueen(int level) {
        if (level == SIZE) {
            printResult();
        } else {
            for (int i = 0; i < SIZE; i++) {
                if (isPossible(level, i)) {
                   queens[level] = i;
                   putQueen(level + 1);
                   queens[level] = EMPTY;
                }
            }
        }
    }

    private boolean isPossible(int level, int cadidate) {
        for (int col = 0; col < level; col++) {
            if (queens[col] == cadidate || level-col == Math.abs(cadidate - queens[col])){
                return false;
            }
        }
        return true;
    }

    private void printResult() {
        System.out.println("RESULT");
        System.out.println(Arrays.toString(queens));
    }


}
